utils/data_utils.py

```diff
@@ -0,0 +1,61 @@
+import math
+
+import numpy as np
+import pandas as pd
+from sklearn.model_selection import train_test_split
+
+
+def preprocessing(df, step: int = 50_000):
```

Comment:
I know this is me being picky, but you could also type the `df` parameter here, like you've done with `step`:

```suggestion
def preprocessing(df: pd.DataFrame, step: int = 50_000):
```